File: L006_Recursion/L006_Recursion.java
import java.io.*;
import java.util.*;

public class L006_Recursion{ 
    //decreasing
    public static void printDecreasing(int n) {
        if (n == 0) return;
        System.out.println(n);
        printDecreasing(n - 1);
    }

    //increasing
    public static void printIncreasing(int n) {
        if (n == 0) return;
        printIncreasing(n - 1);
        System.out.println(n);
   }

    //increasingdecreasing
    public static void pdi(int n) {
        if (n == 0) return;
        System.out.println(n);
        pdi(n - 1);
        System.out.println(n);
    }

   //factorial
    public static int factorial(int n) {
        if (n == 1) return 1;
        int fact = n * factorial(n - 1);
        return fact;
    }

   // power by algebraic
    public static int power(int x, int n) {
        if (n == 0) return 1;
        int num = power(x, n - 1) * x;
        return num;
   }

  //power by log (x^n/2 * x^n/2)
    public static int power(int x, int n) {
        if (n == 0 ) return 1;
        int ans =  power(x,n/2) * power(x,n/2);
        if ( n% 2 == 1) ans *= x;
        return ans;
   }

   //printing pre, in, post zigzag
   public static void pzz(int n) {
    if(n==0){
        return ;
    }
    System.out.print(n + " " );
    pzz(n-1);
    System.out.print(n + " ");
    pzz(n-1);
    System.out.print(n + " ");
   }

   public static void main(String[] args) {
    Scanner scn = new Scanner(System.in);
    int n = scn.next();
    printDecreasing(n);
}
}

File: L006_Recursion/L006_Recursion.cpp
#include <iostream>
using namespace std;

//decreasing
 void printDecreasing(int n){
    // write your code here
    if(n==0) return;
        cout<<n<<endl;
        printDecreasing(n-1);
}

 //increasing
void printIncreasing(int n){
    // write your code here
    if (n == 0) return;
        printIncreasing( n-1);
        cout <<(n)<<endl;
    
}
 
 //increasingdecreasing
void printIncDec(int n){
    // write your code here
    if (n == 0) return;
        cout<< (n) << endl;
         printIncDec( n-1);
        cout<< (n) << endl;

}

//factorial 
int factorial(int n){
    // write your code here
        if(n==1) return 1;
        int fact = n * factorial(n-1);
        return fact;
    
}

// power by algebraic
int power(int x,int n){
        if (n==0) return 1;
        int num = power(x,n-1) * x;
        return num;
}

//power by log (x^n/2 * x^n/2)
int powerLogarithmic(int x,int n){
    // write your code here
        if (n == 0 ) return 1;
        int ans =  powerLogarithmic(x,n/2) ;
        ans *= ans;
        if ( n% 2 == 1) ans *= x;
        return ans;
}

//printing pre, in, post zigzag
void pzz(int n){
    if(n==0){
         return ;
     }
     cout << n << " ";
     pzz(n-1);
     cout <<n << " ";
     pzz(n-1);
     cout <<n << " ";
    }
    
    
// main function 
int main(){
    int x; 
    cout<< "Enter the number :"
    cin >> x;
    int n;
    cout<< "Enter the power of the number :"
    cin >> n;
    cout << powerLogarithmic(x,n);
    return 0;
}
